feat(login): create login and signup form

nappalion opened the pull request:
Implemented the UI for the login and signup form with input for email and password.

**Current Version Demo:** https://imgur.com/fNGQsvI

**RFC:** [https://docs.google.com/document/d/1ixDwuFMW4EJTEnfl-CG6AodOlBydy4RODPEGWv7mMA4/edit?usp=sharing](url)

Changes Made:
- Added "react-native-svg": "12.3.0" to "package.json". (had to run "expo install react-native-svg" to get the working version)
- Added three SVG icons (password visibility and google)
- Created TextInput.js in "components"
- Modified LandingScreen.js in "screens" and Button.js in "components"
- Added unit tests in "Login.test.js" file

nappalion committed:
Create login and signup form

nappalion committed:
Add input validation for login

nappalion committed:
Add password visibility button

nappalion committed:
Add initial jest files for unit testing

nappalion committed:
Fix references to text inputs

nappalion committed:
Change the images to SVG

nappalion committed:
Add unit tests for login screen

nappalion committed:
Removed jest test function

nappalion committed:
Fix missing google login function

lxkedinh commented on packages/app/screens/landing/LandingScreen.js:
```diff
@@ -1,67 +1,296 @@
-      const result = await promptAsync();
+  // skips the code during test
+  if (process.env.NODE_ENV !== 'test') {
+    const [request, response, promptAsync] = Google.useAuthRequest({
+      responseType: 'id_token',
+      expoClientId: '1080245081969-u9lnl9ospj757rq75kiumttqconhnfcc.apps.googleusercontent.com',
+      iosClientId: '1080245081969-u9lnl9ospj757rq75kiumttqconhnfcc.apps.googleusercontent.com',
+      androidClientId: '1080245081969-u9lnl9ospj757rq75kiumttqconhnfcc.apps.googleusercontent.com',
+      webClientId: '1080245081969-u9lnl9ospj757rq75kiumttqconhnfcc.apps.googleusercontent.com',
+    });
```
Are these supposed to be secret credentials? If they are, they should be changed asap if possible and then put in environment variables for the future since the repo's public.

exelarios replied:
Yeah, we should move these into secrets. I believe I was in a rush to migrate from GoogleOAuth into using ExpoAuth Providers.

nappalion replied:
Ok, I'll try and move them to secrets, and I'll fix the yarn/npm problem soon.

lxkedinh reviewed (commented):
Overall, everything looks good to me! Screen looks really nice and runs perfectly on my Galaxy S9+.

nappalion committed:
fix(yarn/npm): remove npm installs and use yarn instead

nappalion committed:
fix(google_secrets): use react native dotenv to move client ID for google auth into secrets

nappalion committed:
fix(app_env): use .env in app workspace instead of server for frontend

lxkedinh committed:
chore: merged new changes from main

nappalion committed:
feat(login/signup): add login functionality and signup screen

nappalion committed:
fix(login_test): Update tests with new imports and navigation

nappalion committed:
feat(landing): Create signup screen

nappalion committed:
fix(signup_screen): Change expo SDK from 46 to 47

ethan-ngo committed:
Added username and password confirmation

ethan-ngo committed:
feat: fist-last name

nappalion committed:
chore(google-login): Add google dependencies and plugin for EAS build

nappalion committed:
fix(google-signin): Fix eas project ID and add signin code

nappalion committed:
fix(google-signin): remove .expo files

nappalion committed:
fix(google-signin): Fix type in idToken

nappalion committed:
Merge branch 'main' of https://github.com/cppsea/icebreak into nipplelion/login_signup_form

nappalion committed:
feat(google-signin): Add local and google signin

nappalion committed:
fix(group-screen): Fix cut off media icons

nappalion committed:
fix(sign-in): Fix double log-in/log-out

nappalion committed:
fix(login-screen): Fix ForgotPasswordScreen and navigation

nappalion committed:
fix(signup): Fix register function on signup screen

nappalion committed:
feat(login-screen): Add auto-login and logging out feature

nappalion committed:
chore(login-screen): Remove comment

lxkedinh commented on packages/app/screens/landing/SignUpScreen.js:
```diff
@@ -0,0 +1,332 @@
+  };
+
+  // State to change the variable with the TextInput
+  const [inputs, setInputs] = React.useState({
```
You're only using the global `React` variable for `useState` in this file, so there's no point in importing it. You can just import useState itself along with the other imports and change your states to just `useState()` instead of `React.useState()`.
```javascript
import { useState, useCallback, useRef } from "react";
```
```javascript
const [inputs, setInputs] = useState({
```

lxkedinh commented on packages/app/screens/landing/SignUpScreen.js:
```diff
@@ -0,0 +1,332 @@
+    const emailRE =
+      /^(([^<>()[\]\.,;:\s@\"]+(\.[^<>()[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i;
+    return email.match(emailRE);
+  };
```
When you declare functions and variables inside React components, you need to think about whether they interact with state or not. This `isValidEmail` utility function doesn't interact with the state of this component (the user's inputs, possible error states). You should put this outside the React component (preferably near the bottom of the file) so this function isn't redefined every time the component rerenders.

lxkedinh commented on packages/app/screens/landing/SignUpScreen.js:
```diff
@@ -0,0 +1,332 @@
+WebBrowser.maybeCompleteAuthSession();
+
+function SignUpScreen({ navigation, route }) {
+  const { user, setUser } = useUserContext();
```
Why do you get the user context here for the signup screen? It should be null at this point since the user is unauthenticated and registering for our app.

lxkedinh commented on packages/app/screens/landing/SignUpScreen.js:
```diff
@@ -0,0 +1,332 @@
+function SignUpScreen({ navigation, route }) {
+  const { user, setUser } = useUserContext();
+
+  const register = async () => {
+    console.log(
+      `Attempting Register with ${inputs.email} and ${inputs.password} at ${ENDPOINT}/auth/local/register`
+    );
+    try {
+      const response = await axios.post(`${ENDPOINT}/auth/local/register`, {
+        firstName: inputs.firstName,
+        lastName: inputs.lastName,
+        email: inputs.email,
+        avatar: "avatar",
+        password: inputs.password,
+      })
+
+      if (response?.data.status == "success") {
+        navigation.navigate("LandingScreen", { email: inputs.email });
+      }
+    } catch (error) {
+      console.log(error.toString());
+    }
+  };
```
The logic in your register function is good here, but you are missing the case where the server finds an existing user with the same email and returns a fail type response.
```JSON
{
  "status": "fail",
  "data": {
    "email": "A user with this email already exists.",
  },
}
```

lxkedinh replied:
We will most likely be implementing this with an alert dialog to show the user when they input an email that is tied to an existing user. I'll have to talk to Napoleon about possibly adding a third-party React Native modal popup/alert dialog library. For now, you can handle this case with a console.log just to make sure it works.

lxkedinh reviewed (changes requested):
Overall really impressive work, Ethan! You have a nice coding style and the signup screen looks and works really nicely so far. There are just a few things we need to fix, but great job so far.

lxkedinh committed:
fix: removed unused babel plugin
- fixed Landing Screen formatting
- updated expo start command in package.json

lxkedinh committed:
fix: fixed events crashing feed screen
- fixed broken ESLint config for frontend
- refactored auto login
- removed babel plugins that caused weird FlatList error

lxkedinh committed:
chore: added cold start frontend yarn script
- allows for starting Expo app with clean bundler cache since bundler cache was causing errors with FeedScreen

lxkedinh committed:
feat: refactored local registration route
- we are moving asking user for first and last name and avatar to onboarding instead of on registration

nappalion committed:
fix(ESLint): Fix problems displayed by ESLint in "app" package
- Added PropTypes dependency for prop checking
- Added "plugin:react/recommended" to fix error on React Native imports
- Add code for logging user out
- Created folder "feed_tabs" for Feed Drawer Tabs

nappalion committed:
fix(login_signup_form): Refactor login/signup code
- Switch from using global React variable to importing functions directly
- Move non-state functions outside React component
- Add check in Signup for already registered email
- Add check in Login for invalid username or password

ethan-ngo committed:
fixed changes in sign up

ethan-ngo committed:
Merge branch 'nipplelion/login_signup_form' of https://github.com/cppsea/icebreak into nipplelion/login_signup_form

nappalion committed:
fix(login): Fixed styling prop to accept arrays and delete tokens if auto-login fails
- Add visible menu Icon in Feed Screen

lxkedinh committed:
fix: incorrect TextInput placeholder prop type
- added check for null tokens to not attempt auto log in

lxkedinh commented on packages/app/screens/landing/SignUpScreen.js:
Remove all code regarding the new user's first and last name since we're moving that to the onboarding screen.

lxkedinh commented on packages/app/screens/landing/SignUpScreen.js:
```diff
@@ -0,0 +1,335 @@
+            testID="googleButton"
+            title="Continue with Google"
+            underlayColor="#ebebeb"
+            onPress={() => {}} //TODO: Implement Google Sign Up
```
Signing up with Google is the same as signing in with Google for our case. We should extract the `handleOnLoginWithGoogle` function and export it to use in both the login and signup screen.

ethan-ngo committed:
Removed name inputs and imported handleGoogle

nappalion committed:
fix(google): Move google login function to utils folder and out of Landing and Signup Screen

lxkedinh committed:
fix: some code cleanup before PR merge
- removed some console.logs
- removed unnecessary passport library after switch to react native google sign in
- fixed root package.json
